torch/_functorch/_aot_autograd/subclass_utils.py

```diff
@@ -111,29 +116,117 @@ def create_subclass_meta(
+                            for (r, is_symint) in zip(size, symint_placeholders)
+                            if is_symint
+                        ]
+                    )
+
+                    # outer_stride
+                    symint_placeholders = [
+                        type(i) is not int for i in meta.outer_stride
+                    ]
+                    assert len(stride) == len(symint_placeholders)
+                    xs_inner.extend(
+                        [
+                            r
+                            for (r, is_symint) in zip(stride, symint_placeholders)
+                            if is_symint
+                        ]
+                    )
+                else:
+                    xs_inner.extend(filter_symints(size))
+                    xs_inner.extend(filter_symints(stride))
+        else:
+            xs_inner.append(x)
+    return xs_inner
+
+
+def unwrap_tensor_subclasses_joint(
+    wrapped_args,
+    *,
+    is_runtime: bool,
+    append_symints: bool,
+    subclass_metas: Optional[List[Union[int, SubclassCreationMeta]]],
+):
+    assert isinstance(wrapped_args, tuple) and len(wrapped_args) == 2
+    primals = unwrap_tensor_subclasses(
+        wrapped_args[0],
+        is_runtime=is_runtime,
+        append_symints=append_symints,
+        subclass_metas=subclass_metas,
+    )
+    tangents = unwrap_tensor_subclasses(
+        wrapped_args[1],
+        is_runtime=is_runtime,
+        append_symints=append_symints,
+        subclass_metas=subclass_metas,
+    )
+    return (primals, tangents)
+
+
+def unwrap_tensor_subclasses_maybe_joint(
+    wrapped_args,
+    *,
+    is_runtime: bool,
+    is_joint_structure: bool,
+    append_symints: bool,
+    subclass_metas: Optional[List[Union[int, SubclassCreationMeta]]],
+):
     if is_joint_structure:
-        assert isinstance(wrapped_args, tuple) and len(wrapped_args) == 2
-        assert isinstance(wrapped_args[0], (tuple, list)) and isinstance(
-            wrapped_args[1], (tuple, list)
+        return unwrap_tensor_subclasses_joint(
+            wrapped_args,
+            is_runtime=is_runtime,
+            append_symints=append_symints,
+            subclass_metas=subclass_metas,
         )
-        unwrapped_args_fw = concat_inner_tensors_from_subclasses(wrapped_args[0])
-        unwrapped_args_tangents = concat_inner_tensors_from_subclasses(wrapped_args[1])
-        unwrapped_args = (unwrapped_args_fw, unwrapped_args_tangents)
     else:
-        assert isinstance(wrapped_args, (list, tuple))
-        unwrapped_args_fw = concat_inner_tensors_from_subclasses(wrapped_args)
-        unwrapped_args = unwrapped_args_fw
-    return unwrapped_args
+        return unwrap_tensor_subclasses(
+            wrapped_args,
+            is_runtime=is_runtime,
+            append_symints=append_symints,
+            subclass_metas=subclass_metas,
+        )
 
 
 def remap_unwrapped_subclass_arg_indices(wrapped_args, static_input_indices):
```

Comment:
@guilhermeleobas I think this function needs to be updated too? Otherwise we will silently compute the wrong indices for which inputs are static, when we desugar subclasses into inner tensors.

(fyi @mlazos, tldr here is that for any graph inputs from dynamo that are subclasses that have dynamic shapes, we now add their SymInt sizes as extra graph args, which makes this remapping a bit more complicated)

It would be good to have a test for this case that would have caught this, Lazos added an existing one for the non-dynamic case here https://github.com/pytorch/pytorch/commit/415d5e53aec33d4ca59ac22b2700cc7f137300b4#diff-c618f2274b6b5ccc533c580549d2e552edbd9fc5ac0da1aa4b00338525c8f78dR1530

Reply:
I added an extra parameter to this test to test dynamic shapes as well.